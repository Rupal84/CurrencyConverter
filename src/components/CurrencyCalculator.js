import React from 'react'
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import Result from './Result'
import { convert } from '../utils/conversionTable'

const handleConversion = (values, actions) => {
  const { fromCurr, toCurr, amount } = values
  const { setStatus } = actions
  const fromCurrency = fromCurr.toUpperCase();
  const toCurrency = toCurr.toUpperCase();
  try {
    const result = convert(fromCurrency, toCurrency, amount);
    const precision = toCurrency === 'JPY' ? 0 : 2;
    setStatus({
      result: result.toFixed(precision),
      resultLabel: `${fromCurrency} to ${toCurrency}`
    });
  } catch (err) {
    setStatus({ error: `Unable to find rate for ${fromCurrency}/${toCurrency}` })
  }
}

function CurrencyCalculator() {
  return (
    <div className="container">
      <h1 className="header">Currency Converter</h1>
      <Formik
        initialValues={{
          fromCurr: '',
          amount: '',
          toCurr: ''
        }}
        validationSchema={Yup.object().shape({
          fromCurr: Yup.string()
            .required('From Currency is required'),
          amount: Yup.number()
            .required('Amount is required'),
          toCurr: Yup.string()
            .required('Email is required')
        })}
        onSubmit={handleConversion}
      >
        {({ errors, status, touched }) => {
          return (
            <Form>
              <div className="row">
                <div className="col-sm-5">
                  <div className="form-group">
                    <label htmlFor={'fromCurr'} className="form-label">From Currency</label>
                    <Field type='text' data-testid="fromCurr"
                      name='fromCurr'
                      className={'form-control' + (errors.fromCurr && touched.fromCurr ? ' is-invalid' : '')}
                    />
                    <ErrorMessage name='fromCurr' component="div" className="invalid-feedback" />
                  </div>
                  <div className="form-group">
                    <label htmlFor={'amount'} className="form-label">Amount</label>
                    <Field type='number' data-testid="amount"
                      name='amount'
                      className={'form-control' + (errors.amount && touched.amount ? ' is-invalid' : '')}
                    />
                    <ErrorMessage name='amount' component="div" className="invalid-feedback" />
                  </div>
                  <div className="form-group">
                    <label htmlFor={'toCurr'} className="form-label">To Currency</label>
                    <Field type='text' data-testid="toCurr"
                      name='toCurr'
                      className={'form-control' + (errors.toCurr && touched.toCurr ? ' is-invalid' : '')}
                    />
                    <ErrorMessage name='toCurr' component="div" className="invalid-feedback" />
                  </div>
                  <div className="btn-group">
                    <span>
                      <button data-testid="submit" type="submit" className="btn btn-success">Convert</button></span>
                    <span><button type="reset" className="btn btn-secondary">Reset</button></span>
                  </div>
                </div>
                <div className="col-sm-6">
                  {status && <Result status={status} />}
                </div>
              </div>
            </Form>
          )
        }
      }
      </Formik></div>
  );
}

export default CurrencyCalculator;
