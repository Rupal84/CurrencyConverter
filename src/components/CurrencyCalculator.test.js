import React from 'react';
import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import CurrencyCalculator from './CurrencyCalculator';
describe('CurrencyCalculator component', () => {
  it('renders correctly', () => {
    const {container} = render(<CurrencyCalculator />);
    expect(container).toMatchSnapshot();
  });
  
  it('fill the form, submit button is clicked for successful conversion', async () => {
    const {getByTestId} = render(<CurrencyCalculator />);
    userEvent.type(getByTestId('fromCurr'),'aud');
    userEvent.type(getByTestId('amount'),'10');
    userEvent.type(getByTestId('toCurr'),'aud');
    await act(async () => userEvent.click(screen.getByRole('button', { name: /convert/i })));
    expect(getByTestId('result')).toHaveTextContent('AUD to AUD: 10.00');
  });

  it('fill the form, submit button is clicked for failed conversion', async () => {
    const {getByTestId} = render(<CurrencyCalculator />);
    userEvent.type(getByTestId('fromCurr'),'xxx');
    userEvent.type(getByTestId('amount'),'10');
    userEvent.type(getByTestId('toCurr'),'aud');
    await act(async () => userEvent.click(screen.getByRole('button', { name: /convert/i })));
    expect(getByTestId('error')).toHaveTextContent('Unable to find rate for XXX/AUD');
  });
})

