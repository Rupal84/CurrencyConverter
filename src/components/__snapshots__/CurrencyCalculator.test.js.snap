// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyCalculator component renders correctly 1`] = `
<div>
  <div
    class="container"
  >
    <h1
      class="header"
    >
      Currency Converter
    </h1>
    <form
      action="#"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-5"
        >
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="fromCurr"
            >
              From Currency
            </label>
            <input
              class="form-control"
              data-testid="fromCurr"
              name="fromCurr"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="amount"
            >
              Amount
            </label>
            <input
              class="form-control"
              data-testid="amount"
              name="amount"
              type="number"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="toCurr"
            >
              To Currency
            </label>
            <input
              class="form-control"
              data-testid="toCurr"
              name="toCurr"
              type="text"
              value=""
            />
          </div>
          <div
            class="btn-group"
          >
            <span>
              <button
                class="btn btn-success"
                data-testid="submit"
                type="submit"
              >
                Convert
              </button>
            </span>
            <span>
              <button
                class="btn btn-secondary"
                type="reset"
              >
                Reset
              </button>
            </span>
          </div>
        </div>
        <div
          class="col-sm-6"
        />
      </div>
    </form>
  </div>
</div>
`;
